File: app/server/backend/routers/habit_checkin.py
"""孕程记 - 好习惯打卡 API。"""

import json
from typing import Optional

from fastapi import APIRouter, Depends, Query
from sqlalchemy import select, func
from sqlalchemy.ext.asyncio import AsyncSession

from backend.database import get_db
from backend.models.habit_checkin import HabitCheckin
from backend.schemas.common import ApiResponse
from backend.schemas.habit_checkin import HabitCheckinCreate, HabitCheckinUpdate, HabitCheckinResponse

router = APIRouter()


@router.post("/habit-checkins", response_model=ApiResponse[HabitCheckinResponse])
async def create_or_update_habit_checkin(data: HabitCheckinCreate, db: AsyncSession = Depends(get_db)):
    """创建/更新好习惯打卡（按日期 upsert）。"""
    result = await db.execute(
        select(HabitCheckin).where(
            HabitCheckin.pregnancy_id == data.pregnancy_id,
            HabitCheckin.date == data.checkin_date,
        )
    )
    existing = result.scalar_one_or_none()

    items_json = json.dumps(data.items, ensure_ascii=False)

    if existing:
        update_data = data.model_dump(exclude_unset=True, exclude={"pregnancy_id", "checkin_date", "items"})
        existing.items = items_json
        for key, value in update_data.items():
            if value is not None:
                setattr(existing, key, value)
        await db.flush()
        return ApiResponse(data=HabitCheckinResponse.model_validate(existing))

    record = HabitCheckin(
        pregnancy_id=data.pregnancy_id,
        date=data.checkin_date,
        items=items_json,
        notes=data.notes,
    )
    db.add(record)
    await db.flush()
    return ApiResponse(data=HabitCheckinResponse.model_validate(record))


@router.get("/habit-checkins", response_model=ApiResponse)
async def list_habit_checkins(
    pregnancy_id: str,
    start_date: Optional[str] = None,
    end_date: Optional[str] = None,
    page: int = 1,
    page_size: int = 20,
    db: AsyncSession = Depends(get_db),
):
    """好习惯打卡列表（分页+日期范围筛选）。"""
    query = select(HabitCheckin).where(HabitCheckin.pregnancy_id == pregnancy_id)
    if start_date:
        query = query.where(HabitCheckin.date >= start_date)
    if end_date:
        query = query.where(HabitCheckin.date <= end_date)
    query = query.order_by(HabitCheckin.date.desc())

    # SQL 层面分页
    count_query = select(func.count()).select_from(query.subquery())
    total = await db.scalar(count_query) or 0
    offset = (page - 1) * page_size
    result = await db.execute(query.offset(offset).limit(page_size))
    records = result.scalars().all()

    return ApiResponse(data={
        "items": [HabitCheckinResponse.model_validate(r) for r in records],
        "total": total, "page": page, "page_size": page_size,
    })


@router.get("/habit-checkins/by-date/{date}", response_model=ApiResponse[HabitCheckinResponse])
async def get_habit_checkin_by_date(
    date: str,
    pregnancy_id: str = Query(...),
    db: AsyncSession = Depends(get_db),
):
    """获取指定日期的好习惯打卡。"""
    result = await db.execute(
        select(HabitCheckin).where(
            HabitCheckin.pregnancy_id == pregnancy_id,
            HabitCheckin.date == date,
        )
    )
    record = result.scalar_one_or_none()
    if not record:
        return ApiResponse(code=1001, data=None, message="该日期无打卡记录")
    return ApiResponse(data=HabitCheckinResponse.model_validate(record))


@router.get("/habit-checkins/{checkin_id}", response_model=ApiResponse[HabitCheckinResponse])
async def get_habit_checkin(checkin_id: str, db: AsyncSession = Depends(get_db)):
    """好习惯打卡详情。"""
    record = await db.get(HabitCheckin, checkin_id)
    if not record:
        return ApiResponse(code=1001, data=None, message="记录不存在")
    return ApiResponse(data=HabitCheckinResponse.model_validate(record))


@router.put("/habit-checkins/{checkin_id}", response_model=ApiResponse[HabitCheckinResponse])
async def update_habit_checkin(checkin_id: str, data: HabitCheckinUpdate, db: AsyncSession = Depends(get_db)):
    """更新好习惯打卡。"""
    record = await db.get(HabitCheckin, checkin_id)
    if not record:
        return ApiResponse(code=1001, data=None, message="记录不存在")
    if data.items is not None:
        record.items = json.dumps(data.items, ensure_ascii=False)
    if data.notes is not None:
        record.notes = data.notes
    await db.flush()
    return ApiResponse(data=HabitCheckinResponse.model_validate(record))


@router.delete("/habit-checkins/{checkin_id}", response_model=ApiResponse)
async def delete_habit_checkin(checkin_id: str, db: AsyncSession = Depends(get_db)):
    """删除好习惯打卡。"""
    record = await db.get(HabitCheckin, checkin_id)
    if not record:
        return ApiResponse(code=1001, data=None, message="记录不存在")
    await db.delete(record)
    return ApiResponse(message="删除成功")

File: server/node/routes/habit-checkin.js
const express = require('express');
const router = express.Router();
const db = require('../db');
const config = require('../config');

router.post('/habit-checkins', async (req, res) => {
  try {
    const { pregnancy_id, date, items, note } = req.body;
    if (!pregnancy_id || !date) {
      return res.json({ code: 1001, data: null, message: '缺少必要参数pregnancy_id或date' });
    }

    const existing = await db.queryOne(
      'SELECT id FROM habit_checkin WHERE pregnancy_id = ? AND date = ?',
      [pregnancy_id, date]
    );

    const itemsJson = JSON.stringify(items || []);

    if (existing) {
      await db.run(
        'UPDATE habit_checkin SET items = ?, notes = ?, updated_at = CURRENT_TIMESTAMP WHERE id = ?',
        [itemsJson, note || '', existing.id]
      );
      res.json({ code: 0, data: { id: existing.id, updated: true }, message: '更新成功' });
    } else {
      const id = db.generateId();
      await db.run(
        'INSERT INTO habit_checkin (id, pregnancy_id, date, items, notes, created_at, updated_at) VALUES (?, ?, ?, ?, ?, CURRENT_TIMESTAMP, CURRENT_TIMESTAMP)',
        [id, pregnancy_id, date, itemsJson, note || '']
      );
      res.json({ code: 0, data: { id, created: true }, message: '创建成功' });
    }
  } catch (error) {
    res.json({ code: 1001, data: null, message: error.message });
  }
});

router.get('/habit-checkins', async (req, res) => {
  try {
    const { pregnancy_id, start_date, end_date, page = 1, page_size = 20 } = req.query;
    if (!pregnancy_id) {
      return res.json({ code: 1001, data: null, message: '缺少pregnancy_id参数' });
    }

    const offset = (parseInt(page) - 1) * parseInt(page_size);
    let whereClause = 'WHERE pregnancy_id = ?';
    let params = [pregnancy_id];

    if (start_date) {
      whereClause += ' AND date >= ?';
      params.push(start_date);
    }
    if (end_date) {
      whereClause += ' AND date <= ?';
      params.push(end_date);
    }

    const totalResult = await db.queryOne(
      `SELECT COUNT(*) as total FROM habit_checkin ${whereClause}`, params
    );
    const total = totalResult.total;

    const list = await db.queryAll(
      `SELECT * FROM habit_checkin ${whereClause} ORDER BY date DESC LIMIT ? OFFSET ?`,
      [...params, parseInt(page_size), offset]
    );

    list.forEach(item => {
      if (typeof item.items === 'string') {
        try { item.items = JSON.parse(item.items); } catch (e) { item.items = []; }
      }
    });

    res.json({
      code: 0,
      data: {
        list,
        pagination: {
          page: parseInt(page),
          page_size: parseInt(page_size),
          total,
          total_pages: Math.ceil(total / parseInt(page_size))
        }
      },
      message: 'success'
    });
  } catch (error) {
    res.json({ code: 1001, data: null, message: error.message });
  }
});

router.get('/habit-checkins/by-date/:date', async (req, res) => {
  try {
    const { date } = req.params;
    const { pregnancy_id } = req.query;

    if (!pregnancy_id) {
      return res.json({ code: 1001, data: null, message: '缺少pregnancy_id参数' });
    }

    const record = await db.queryOne(
      'SELECT * FROM habit_checkin WHERE pregnancy_id = ? AND date = ?',
      [pregnancy_id, date]
    );

    if (record && typeof record.items === 'string') {
      try { record.items = JSON.parse(record.items); } catch (e) { record.items = []; }
    }

    res.json({ code: 0, data: record || null, message: 'success' });
  } catch (error) {
    res.json({ code: 1001, data: null, message: error.message });
  }
});

router.get('/habit-checkins/:id', async (req, res) => {
  try {
    const { id } = req.params;
    const record = await db.queryOne('SELECT * FROM habit_checkin WHERE id = ?', [id]);

    if (!record) {
      return res.json({ code: 1001, data: null, message: '记录不存在' });
    }

    if (typeof record.items === 'string') {
      try { record.items = JSON.parse(record.items); } catch (e) { record.items = []; }
    }

    res.json({ code: 0, data: record, message: 'success' });
  } catch (error) {
    res.json({ code: 1001, data: null, message: error.message });
  }
});

router.put('/habit-checkins/:id', async (req, res) => {
  try {
    const { id } = req.params;
    const { items, note } = req.body;

    const existing = await db.queryOne('SELECT id FROM habit_checkin WHERE id = ?', [id]);
    if (!existing) {
      return res.json({ code: 1001, data: null, message: '记录不存在' });
    }

    const updates = [];
    const params = [];

    if (items !== undefined) {
      updates.push('items = ?');
      params.push(JSON.stringify(items));
    }
    if (note !== undefined) {
      updates.push('note = ?');
      params.push(note);
    }

    updates.push('updated_at = CURRENT_TIMESTAMP');
    params.push(id);

    await db.run(`UPDATE habit_checkin SET ${updates.join(', ')} WHERE id = ?`, params);

    res.json({ code: 0, data: { id, updated: true }, message: '更新成功' });
  } catch (error) {
    res.json({ code: 1001, data: null, message: error.message });
  }
});

router.delete('/habit-checkins/:id', async (req, res) => {
  try {
    const { id } = req.params;

    const existing = await db.queryOne('SELECT id FROM habit_checkin WHERE id = ?', [id]);
    if (!existing) {
      return res.json({ code: 1001, data: null, message: '记录不存在' });
    }

    await db.run('DELETE FROM habit_checkin WHERE id = ?', [id]);

    res.json({ code: 0, data: { id, deleted: true }, message: '删除成功' });
  } catch (error) {
    res.json({ code: 1001, data: null, message: error.message });
  }
});

module.exports = router;
